### PR TITLE
Convert average routines to C++

### PR DESCRIPTION
**Pull request by harpolea**
This converts the routines in `average.F90` to C++. 

I've tested it with MPI and OpenMP for reacting bubble and wdconvect, and I don't think it should produce any changes in the output. 

I'm not entirely sure the GPU reduction is 100% correct, so while it runs I'd appreciate if someone could read over that to make sure it looks correct.

### CODE REVIEW

**Comment by doreenfan** [comment]
Spherical wdconvect MPI+OMP tests pass with zero output diffs, but reacting bubble 2d fails as well as 3d with OMP with diffs larger than roundoff errors. 
I've narrowed it down to the diffs coming from converting `average()` into C++, but that particular function is quite simple and straightforward. I don't see any obvious bugs, so perhaps the errors are simply built over the many calls to average() in the code.

**Comment by harpolea** [comment]
Removing the tiling may make it work for OMP, however this will break it on the GPU, so something else needs to be done to fix the reduction across threads for both

**Comment by zingale** [comment]
test results are roundoff (aside from conductivity constants changes)
http://groot.astro.sunysb.edu/MAESTROeX/test-suite/gfortran/2020-03-16-001/index.html

File: Source/MaestroMakeBeta0.cpp [hunk 1]
```diff
@@ -82,13 +84,18 @@ Maestro::MakeBeta0(RealVector& beta0,
...
                             // piecewise linear reconstruction of rho0,
                             // gamma1bar, and p0 -- see paper III, appendix C
                             Real del = 0.5 * (rho0[n+max_lev*(r+1)] - rho0[n+max_lev*(r-1)])/drc;
                             Real dpls = 2.0 * (rho0[n+max_lev*(r+1)] - rho0[n+max_lev*r])/drp;
                             Real dmin = 2.0 * (rho0[n+max_lev*r] - rho0[n+max_lev*(r-1)])/drm;
                             Real slim = min(fabs(dpls), fabs(dmin));
+                            // Print() << "slim = " << slim << " del = " << del << " dpls = " << dpls << " dmin = " << dmin << std::endl;
+                            // Print() << "rho0 = " << rho0[n+max_lev*(r-1)] << " " << rho0[n+max_lev*r] << " " << rho0[n+max_lev*(r+1)] << std::endl;
+                            // Print() << "drc, drp, drm = " << drc << " " << drp <<  " " << drm << std::endl;
+                            slim = slim == slim ? slim : 0.0;
```
Comment (zingale) [inline Source/MaestroMakeBeta0.cpp]: what does this do?

**Pull request merged**